### ISSUE TITLE
Package request: Tube archivist

### ISSUE THREAD

**Issue by hufforguk** [NixOS/nixpkgs#248397]
**Project description**

Once your YouTube video collection grows, it becomes hard to search and find a specific video. That's where Tube Archivist comes in: By indexing your video collection with metadata from YouTube, you can organize, search and enjoy your archived YouTube videos without hassle offline through a convenient web interface. 

**Metadata**

* homepage URL: https://www.tubearchivist.com/
* source URL: https://github.com/tubearchivist/tubearchivist
* license: gpl3
* platforms: linux

**Comment by EricTheMagician** [comment]
added a missing tubearchivist dependency:
https://github.com/NixOS/nixpkgs/pull/322094

**Comment by 06kellyjac** [comment]
One blocker for a tubearchivist module on nixos was it wants Redis with RedisJSON. I don't believe we provide redis modules in nixpkgs right now.

v0.5.0 which is coming (subject to change) doesn't depend on RedisJSON.
https://gist.github.com/bbilly1/7b6abc52ab689f56671bf9011879379a

---

personally I'm running tubearchivist, upstream redis, and the tubearchivist-es image for elastic via `virtualisation.oci-containers.containers` (with podman as the backend)
Had some networking and user permissions hoops to jump through but running fine so far :)

**Comment by 06kellyjac** [comment]
Sorry for the delay. Here's my config with v0.5.0 (which doesn't need RedisJSON anymore)

```nix
{ config, pkgs, ... }:
let
  tubearchivistRedisConfig = config.services.redis.servers.tubearchivist;
in
{
  users.users.tubearchivist = {
    isSystemUser = true;
    uid = 985;
  };
  users.users.tubearchivist.group = "share";

  virtualisation.oci-containers.backend = "podman";
  virtualisation.podman.defaultNetwork.settings.dns_enabled = true;
  virtualisation.oci-containers.containers.tubearchivist = {
    image = "bbilly1/tubearchivist:v0.5.0";
    environment = {
      ES_URL = "http://127.0.0.1:9200";
      REDIS_CON = "redis://${tubearchivistRedisConfig.bind}:${builtins.toString tubearchivistRedisConfig.port}";
      HOST_UID = "${builtins.toString config.users.users.tubearchivist.uid}";
      HOST_GID = "${builtins.toString config.users.groups.share.gid}";
      TA_USERNAME = "tubearchivist";
      TA_PASSWORD = "verysecret";
      ELASTIC_PASSWORD = "verysecret";
      TA_HOST = "https://youraddress";
      TZ = config.time.timeZone;
    };
    volumes = [
      "/some/dir/tubearchivist:/youtube"
      "/var/cache/tubearchivist:/cache"
    ];
    extraOptions = [ "--network=host" ];
    # ports = [ "${builtins.toString port}:8000" ];
  };
  networking.firewall.allowedTCPPorts = [ 8000 ];

  # need elastic api reporting v8
  virtualisation.oci-containers.containers.tubearchivist-es = {
    image = "bbilly1/tubearchivist-es";
    environment = {
      ELASTIC_PASSWORD = "verysecret";
      ES_JAVA_OPTS = "-Xms1g -Xmx1g";
      "xpack.security.enabled" = "true";
      "discovery.type" = "single-node";
      "path.repo" = "/usr/share/elasticsearch/data/snapshot";
    };
    volumes = [ "/var/lib/tubearchivist/elasticsearch:/usr/share/elasticsearch/data" ];
    ports = [ "9200:9200" ];
  };

  services.redis.servers.tubearchivist.enable = true;
  services.redis.servers.tubearchivist.port = 6379;
}
```

Possible improvements:

* Play around with networking again to avoid using host.
* use `virtualisation.oci-containers.containers.<name>.environmentFiles` to set `ELASTIC_PASSWORD` for both containers outside of nix store
  * Similar for `TA_USERNAME` and `TA_PASSWORD`, I need to double check if they're used after you set your own users or just move it to secret env file anyway
* make tubearchivist and elastic run as non-root, might not bother if we just package it natively

The fact there was a manual migration required for the breaking change might suggest holding off on packaging until a v1 stabilizes

**Comment by z8512** [comment]
Thank u

**Comment by ibizaman** [comment]
Looks like redisjson requirement has been removed https://github.com/tubearchivist/tubearchivist/commit/fbd7c19a93e64225f909c336396fe9cb3f20e13f

**Comment by 06kellyjac** [comment]
Yeah, the config above uses native nixos redis rather than a redis container with redisjson.
The upstream project shows some concerns with Elastic Search version mismatches but it's a lot easier not needing redisjson.

**Comment by 06kellyjac** [comment]
Loading secrets from files natively would be nice but we can still read them at runtime ourselves.
E.g. `./nixos/modules/services/misc/renovate.nix`


The upstream GH isn't taking new FR tickets but it's something to maybe discuss in the Discord. Or a relatively simple change to just make a PR (or maintain as a patch only because it's basic and that file probably wont change much).

**Comment by Skyb0rg007** [comment]
I think a patch is a decent option. We need to patch the source anyways to move the `staticfiles` path outside the source directory, which it sounds like I'd have a hard time upstreaming.

Unfortunately I also realized that Elasticsearch 8.x is required, which has been blocked from Nixpkgs for over a year (#324787). If someone else gets that merged then I'll create the PR for the Tube Archivist package + module.

Also, the issue with secrets isn't how to get them into Tube Archivist. The difficulty is configuring Elasticsearch to have a user + password, which is done in the Dockerfile but is hard to convert into a systemd unit since a NixOS elasticsearch instance could be used for other (non-tubearchivist) applications.
